code, tests & examples for (i) converting XML <-> JSON, & (ii) residuals

matthewjohnpayne opened the pull request:
The changes in this branch are to perform the following

(1) allow conversion between XML <-> JSON files. 
 - The purpose is for my own convenience elsewhere in some other MPC development tasks, and is not specifically intended to be used to allow users to submit data in JSON format. It does however open the possibility of allowing that in the future. 
 - As per the extant XML <-> PSV conversion routines, these new XML <-> JSON routines are pure *conversion* routines: they do NOT perform any *validation* on the XML or JSON, nor do they require that the input be valid. 

(2) provide examples (and validation tests) of ADES files that include RESIDUALS. 
 - It was not clear to me what the hell the allowed format was supposed to be, nor that there are two allowed variations-on-a-theme (one inside \<optical\>, the other 'standalone'). As such I have made a couple of example data files (new_tests/input/*Residual.xml) , and provided some quick validation tests (new_tests/test_validationResiduals.py) to reassure myself that they work.

federicaspoto commented on new_tests/input/obsResidual.xml:
```diff
@@ -0,0 +1,107 @@
+                        <orbProd>Someone</orbProd>
+                        <orbID>1</orbID>
+        <!-- Starting OpticalRes sub-sub-block -->
+                                  <resRA>1.2</resRA>
+                                  <resDec>1.2</resDec>
+                                  <selAst>A</selAst>
+                                  <sigRA>1.2</sigRA>
+                                  <sigDec>1.2</sigDec>
+                                  <sigCorr>0.5</sigCorr><!-- optional -->
+                                  <sigTime>0.5</sigTime><!-- optional -->
+                                  <biasRA>1.2</biasRA><!-- optional -->
+                                  <biasDec>1.2</biasDec><!-- optional -->
+                                  <biasTime>1.2</biasTime><!-- optional -->
+        <!-- Starting OpticalResMag sub-sub-block -->
+                                  <photProd>Someone else</photProd><!-- optional -->
+                                  <resMag>1.2</resMag>
+                                  <selPhot>A</selPhot>
+                                  <sigMag>1.2</sigMag>
+                                  <biasMag>1.2</biasMag>
+                                  <photMod>abc123</photMod>
```
Not a big problem, but the formatting looks weird to me. Is this how it's supposed to look like?

federicaspoto commented on new_tests/input/standaloneResidual.xml:
```diff
@@ -0,0 +1,28 @@
+        <!-- Starting OpticalResidual sub-block -->
+                        <orbProd>Someone</orbProd>
+                        <orbID>1</orbID>
+        <!-- Starting OpticalRes sub-sub-block -->
+                                  <resRA>1.2</resRA>
+                                  <resDec>1.2</resDec>
+                                  <selAst>A</selAst>
+                                  <sigRA>1.2</sigRA>
+                                  <sigDec>1.2</sigDec>
+                                  <sigCorr>0.5</sigCorr><!-- optional -->
+                                  <sigTime>0.5</sigTime><!-- optional -->
+                                  <biasRA>1.2</biasRA><!-- optional -->
+                                  <biasDec>1.2</biasDec><!-- optional -->
+                                  <biasTime>1.2</biasTime><!-- optional -->
+        <!-- Starting OpticalResMag sub-sub-block -->
+                                  <photProd>Someone else</photProd><!-- optional -->
+                                  <resMag>1.2</resMag>
+                                  <selPhot>A</selPhot>
+                                  <sigMag>1.2</sigMag>
+                                  <biasMag>1.2</biasMag>
+                                  <photMod>abc123</photMod>
```
Same questions about formatting

federicaspoto reviewed (approved):
@matthewjohnpayne looks good to me. I added a couple of comments. 
Also, could you please add a couple of tests about using json2xml as a routine and not calling it using subprocess? I believe all the other tests show the different options. 
Thanks!

matthewjohnpayne commented:
@federicaspoto : Thanks for looking at this. 
 - I added tests that import the json2xml & xml2json routines and call them as functions (in addition to the extant command-line call tests). 
 - I also improved all of the tests such that pytest can be called from anywhere (as previously written, I'd inadvertently  required that the user run the tests from within new_tests) 

With regards to your comment/questions regarding the format of `input/*Residual.xml`:
 - The parts that looks like `\<!-- ***** -- \>` are the standard way to add *comments * to XML files. 
 - I'm primarily using them here as a means to highlight to my future-self that certain of the fields in the sample input residuals are *optional* (while the rest are required by the ADES schema).

federicaspoto commented:
@matthewjohnpayne thank you! 
About my comment on the formatting, it was about the fields themselves, not the comments. But it is not important. It just looked weird on my screen.

matthewjohnpayne commented:
@federicaspoto : no worries. 
I'm going to leave this open and let @stevechesley have a chance to look at it before I merge anything.

stevechesley reviewed (approved):
I just added shebangs to the new python scripts. I did not add the `from __future__ import` stuff, but perhaps it should be there? Not really sure what that does...

Should new scripts be renamed, e.g., from `xml2json.py` to `xmltojson.py`, for consistency with other scripts in the repo? Seems easier to follow the same naming convention. 

Also, I had to pip install pyyaml and xmltodict to get these new scripts to work. Should they be added to the Python/requirements.txt file?

matthewjohnpayne commented:
@stevechesley : thanks for adding in the shebangs. 
I added the pyyaml and xmltodict to the requirements file. 
I renamed the code to be 'xmltojson', etc
For what it's worth, I also added in the __future__ stuff, although I've no idea if there's any value to it anymore. 
All my tests still pass, so I'm going to go ahead and merge.
